go-controller/pkg/metrics/master.go

```diff
@@ -681,6 +713,374 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	// This time is stored for later processing and is used as the start time.
+	// An operation is returned to the caller, which they can bundle with their existing transactions sent to OVN which
+	// will tell OVN to measure how long it takes to configure all nodes with the config in the transaction.
+	// Config duration then waits for OVN to configure all nodes and calculates the time delta.
+	// The measurements are limited and are only allowed during certain time periods to reduce the stress on OVN.
+
+	// ** configuration duration monitor - caveats **
+	// The duration described above does not give you an exact time duration for how long it takes to configure your
+	// k8 kind. When you are recording how long it takes OVN to complete your configuration to all nodes, other
+	// transactions may have occurred which may increases the overall time. You may also get longer processing times if one
+	// or more nodes are unavailable because we are measuring how long the functionality takes to apply to ALL nodes.
+
+	// ** configuration duration monitor - How the duration of the config is measured within OVN **
+	// We increment the nb_cfg integer value in the NB_Global table.
+	// ovn-northd notices the nb_cfg change and copies the nb_cfg value to SB_Global table field nb_cfg along with any
+	// other configuration that is changed in OVN Northbound database.
+	// All ovn-controllers detect nb_cfg value change and generate a 'barrier' on the openflow connection to the
+	// nodes ovs-vswitchd. Once ovn-controllers receive the 'barrier processed' reply from ovs-vswitchd which
+	// indicates that all relevant openflow operations associated with NB_Globals nb_cfg value have been
+	// propagated to the nodes OVS, it copies the SB_Global nb_cfg value to its Chassis_Private table nb_cfg record.
+	// ovn-northd detects changes to the Chassis_Private startRecords and computes the minimum nb_cfg for all Chassis_Private
+	// nb_cfg and stores this in NB_Global hv_cfg field along with a timestamp to field hv_cfg_timestamp which
+	// reflects the time when the slowest chassis catches up with the northbound configuration.
+
+	// create a map with key as the name of the k8 kind you want to measure and value with the histogram to report
+	// the duration to configure the kind to all nodes. Suggestion to name the key like so: {k8 kind}. E.g. pod
+	cdr.reportKind = make(map[string]prometheus.Histogram)
+	// register metrics which will take account of configuration start time and the time it
+	// takes OVN to configure all nodes. Note that the result will only give an upper bound limit of how long it
+	// takes to roll out the configuration to all nodes. To make matters more uncertain, the measurements are currently
+	// sampled.
+	cdr.reportKind[PodKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_pod_duration_seconds",
```

Comment:
I realize this is somewhat of a large change this late in the game, but a thought occurs to me: it would be a bit more idiomatic for this to be a single metric, with the Kind indicated by a label. Then we can aggregate on general latency, or drill down in to specific Kinds as necessary.

That would also suggest making the PodRecorder generic across object types.

Reply:
yeah, that would be nicer. Its not too late as I am going to wait for #2909 as it seems more important than this. Thank you.

Reply:
@squeed I just realised I would have to give up using buckets if I want to use labels.
I believe buckets do not support labelling. 
I can use a gauge or count vector.

Reply:
@martinkennelly https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewHistogramVec should do the job...

Reply:
I am an idiot.. Thanks casey